sklearn/decomposition/nmf.py

```diff
@@ -254,22 +344,287 @@ def _nls_subproblem(V, W, H, tol, max_iter, sigma=0.01, beta=0.1):
+        The regularization mixing parameter, with 0 <= l1_ratio <= 1.
+        For l1_ratio = 0 the penalty is an L2 penalty.
+        For l1_ratio = 1 it is an L1 penalty.
+        For 0 < l1_ratio < 1, the penalty is a combination of L1 and L2.
+
+    regularization : 'both' | 'components' | 'transformation' | None
+        Select whether the regularization affects the components (H), the
+        transformation (W), both or none of them.
+
+    update_H : boolean, default: True
+        Set to True, both W and H will be estimated from initial guesses.
+        Set to False, only W will be estimated.
+
+    verbose : integer, default: 0
+        The verbosity level.
+
+    shuffle : boolean, default: False
+        If True, the samples will be taken in shuffled order during
+        coordinate descent.
+
+    random_state : integer seed, RandomState instance, or None (default)
+        Random number generator seed control.
+
+    Returns
+    -------
+    W : array-like, shape (n_samples, n_components)
+        Solution to the non-negative least squares problem.
+
+    H : array-like, shape (n_components, n_features)
+        Solution to the non-negative least squares problem.
+
+    n_iter : int
+        The number of iterations done by the algorithm.
 
-    init :  'nndsvd' |  'nndsvda' | 'nndsvdar' | 'random'
+    References
+    ----------
+    Hsieh, Cho-Jui, and Inderjit S. Dhillon. "Fast coordinate descent methods
+    with variable selection for non-negative matrix factorization."
+    Proceedings of the 17th ACM SIGKDD. ACM, 2011.
+    """
+    # so W and Ht are both in C order in memory
+    Ht = check_array(H.T, order='C')
+    X = check_array(X, accept_sparse='csr')
+
+    alpha_H = 0.
+    alpha_W = 0.
+    if regularization in ('both', 'components'):
+        alpha_H = alpha
+    if regularization in ('both', 'transformation'):
+        alpha_W = alpha
+
+    rng = check_random_state(random_state)
+
+    for n_iter in range(max_iter):
+            v = 0.
+
+            # Update W
+            v += _update_coordinate_descent(X, W, Ht, alpha_W, l1_ratio,
+                                            shuffle, rng)
+            # Update H
+            if update_H:
+                v += _update_coordinate_descent(X.T, Ht, W, alpha_H, l1_ratio,
+                                                shuffle, rng)
+
+            if n_iter == 0:
+                violation_init = v
+
+            if violation_init == 0:
+                break
+
+            if verbose:
+                print("violation:", v / violation_init)
+
+            if v / violation_init <= tol:
+                if verbose:
+                    print("Converged at iteration", n_iter + 1)
+                break
+
+    return W, Ht.T, n_iter
+
+
+def non_negative_matrix_factorization(X, W=None, H=None, n_components=None,
+                                      init='random', update_H=True,
+                                      solver='proj-grad', tol=1e-4,
```

Comment:
The default solver is not consistent with that of the class. I would use coordinate descent here.